doc/sphinx/proof-engine/ltac.rst

```diff
@@ -28,159 +29,92 @@ especially about its foundations, please refer to :cite:`Del00`.
-                    : `qualid` [`ident` ... `ident`] ::= `ltac_expr`
+The grammar uses multiple :n:`ltac_expr*` nonterminals to express how subexpressions
+are grouped when they're not fully parenthesized.  For example, in many programming
+languages, `a*b+c` is interpreted as `(a*b)+c` because `*` has
+higher precedence than `+`.  Usually `a/b/c` is given the :gdef:`left associative`
+interpretation `(a/b)/c` rather than the :gdef:`right associative` interpretation
+`a/(b/c)`.
 
-Tactics in terms
-~~~~~~~~~~~~~~~~
+In Coq, the expression :n:`try repeat @tactic__1 || @tactic__2; @tactic__3; @tactic__4`
+is interpreted as :n:`(try (repeat (@tactic__1 || @tactic__2)); @tactic__3); @tactic__4`
+because `||` is part of :token:`ltac_expr2`, which has higher precedence than
+:tacn:`try` and :tacn:`repeat` (at the level of :token:`ltac_expr3`), which
+in turn have higher precedence than `;`, which is part of :token:`ltac_expr4`.
+(A *lower* number in the nonterminal name means *higher* precedence in this grammar.)
 
-.. insertprodn term_ltac term_ltac
+The constructs in :token:`ltac_expr` are :term:`left associative`.
+
+.. insertprodn ltac_expr tactic_atom
 
 .. prodn::
-   term_ltac ::= ltac : ( @ltac_expr )
+   ltac_expr ::= {| @ltac_expr4 | @binder_tactic }
+   ltac_expr4 ::= @ltac_expr3 ; {| @ltac_expr3 | @binder_tactic }
+   | @ltac_expr3 ; [ @for_each_goal ]
+   | @ltac_expr3
+   ltac_expr3 ::= @l3_tactics
+   | @ltac_expr2
+   ltac_expr2 ::= @ltac_expr + {| @ltac_expr2 | @binder_tactic }
+   | @ltac_expr || {| @ltac_expr2 | @binder_tactic }
+   | @l2_tactics
+   | @ltac_expr1
+   ltac_expr1 ::= @value_tactic
+   | @qualid {* @tactic_arg }
+   | @l1_tactics
+   | @ltac_expr0
+   tactic_arg ::= @value_tactic
+   | @term
+   | ()
+   ltac_expr0 ::= ( @ltac_expr )
+   | [> @for_each_goal ]
+   | @tactic_atom
+   tactic_atom ::= @int
+   | @qualid
+   | ()
+
+.. todo PR For the moment, I've left the language constructs like +, || and ; unchanged in the grammar.
+   Not sure what to do with them.  If we just make these indirections I think the grammar no longer
+   gives you an overall idea of the concrete grammar without following the hyperlinks for many terms--not so easy
+   (e.g. I have a construct and I want to figure out which productions generate it so I can read about them).
+   We should think about eventually having a cheat sheet for the constructs, perhaps as part of the
+   chapter introduction (use case: I know there's a construct but I can't remember its syntax).  They
+   do show up in the index but they're not so easy to find.  I had thought a little about putting
+   an ltac expression cheat sheet at the top of the tactics index.  Unconventional, but people would
+   see it and remember how to find it.
+
+   OTOH, as you rightly note, they are not really tactics.  Looking for better ideas that we are OK with.
+
+.. note::
+
+   Syntactically, almost all tactics are in :token:`l3_tactics`.
```

Comment:
`simple_tactic` is in `l1_tactics`.

Reply:
OK, but then the following becomes confusing:

>The documentation also describes the following as tactics because they can generally be used anywhere a tactic can even though they fit into other levels of ltac_expr:
>- ...
>- l1_tactics are: first, solve, idtac, fail and gfail as well as match, match goal and their lazymatch and multimatch variants.

If the wording could be tweaked to make it a tad less confusing, it would be better. Not a big deal though.